### ISSUE TITLE
Allow strict compiler flags (-Wall -Wextra -Werror -pedantic)

### ISSUE THREAD

**Issue by anderejd** [Dav1dde/glad#44]
I will begin by saying thank you for this project! :)

Using -pedantic with GCC (tested with 5.3.0) will result in a lot of the following warnings (or errors for us using -Werror):
**ISO C forbids conversion of object pointer to function pointer type**

I solved it by modifying my generated glad.h:

```
typedef void (*GLADvoidfn)(void);
typedef GLADvoidfn (*GLADloadproc)(const char *name);

```

This will avoid the casts from pointer-to-void to pointer-to-function inside glad.

In my calling code I also needed to add this since I use SDL which also uses pointers-to-void for function pointers:

```
static GLADvoidfn
getprocaddress_wrapper(const char *name)
{
#pragma GCC diagnostic push
#pragma GCC diagnostic ignored "-pedantic"
    return (GLADvoidfn) SDL_GL_GetProcAddress(name);
#pragma GCC diagnostic pop
}
```

I have not tried this on other compilers yet.
Would you like a pull request?

**Comment by Dav1dde** [comment]
Hey, thanks for the report. I could reproduce the problem and looked further into the issue. It turns out ISO C does not guarantee casts from `void*` to function pointers to succeed, since the standard assumes `void*` is a different type (data) than a function pointer and these can have different representations. [Steve Summit explains that better than me, in his Stackoverflow answer](http://stackoverflow.com/a/31527867/969534).

There are workarounds to trick the compiler (see SO answer), or your solution which only shifts the problem to the loader function. 

The only real solution is to compile the relevant files (glad) without `-pedantic`, simply because casting from `void*` to a function pointer is not ISO C semantics.

**Comment by rajder** [comment]
Hi!

Yes, I searched for the standard info too and found the same thing, void pointers are pointers to object and casting to function pointer is implementation defined. (however Posix does define the cast, but that is still not in the ISO standard)

I'm not sure that I agree with your conclusion, the implementation defined behavior is still there even if one would turn off the warnings in the compiler, so I would not call that a solution.

My thinking is that moving the necessary cast (at least working with SDL) outside glad would at least keep glad compliant to the standard and leave the decision about lowering warning levels vs tricking the compiler to be silent up to the user.

Best Regards /Anders

**Comment by Dav1dde** [comment]
I think you are right, changing the return type of the loader might just work, if you want you can make a pull request.

But there is a new issue, the 'builtin' loader currently returns `void*`, one solution would be to 'trick' the compiler (the trick was apparently removed from the dlsym POSIX docs), which I don't like, or to not support compiling with `-pedantic` (the current situation). Input welcome.

**Comment by rajder** [comment]
Ok cool, I could possibly find some time to put together a pull request in the weekend.
I will look into the built-in loader too before sending a pull request.

**Comment by rajder** [comment]
I just realised that I overcomplicated my proposed fix.

After changing the signature of GLADloadproc I can correctly cast SDL_GL_GetProcAddress with -pedantic, no need for the #pragma GCC diagnostic shenanigans.

However, the cast is correct, but I am not sure if the function call through the pointer with the new type is correct/defined... The result if using SDL is calling `void* SDL_GL_GetProcAddress(const char* proc)` through a pointer of type:

```
typedef void (*GLADvoidfn)(void);
typedef GLADvoidfn (*GLADloadproc)(const char *name);

or

void (*(*GLADloadproc)(const char *))(void) 

```

What do you think?
Maybe you were right all along and the only correct way to do it is to lower warning levels for glad?

**Comment by Dav1dde** [comment]
The root issue (as I understand it) is: casting a pointer (which in ISO C is a pointer to an object) to a function(-pointer) is not defined. *But* dlsym does it and on every platform where dlsym exists, the cast is defined and works, on all platforms where this does not work (are there any?), there is also no way OpenGL will work, since it does the loading through void* as well. 
So we don't need to care about the cast itself, only about the warning beeing 'annoying'.

I still think the "correct" solution is simply to drop `-pedantic` for glad. But nothing speaks against making it easier for the end user. So as long as the changes make it easier and don't break existing code I am fine with them.

So the cast from `glfwGetProcAddress` to `GLADloadproc` still needs to succeed (same for SDL), `gladLoadGLLoader((GLADloadproc) glfwGetProcAddress)`, I am pretty sure doing your initial change:
```c
typedef void (*GLADvoidfn)(void);
typedef GLADvoidfn (*GLADloadproc)(const char *name);
```
Can't 100% confirm if that works with `-ansi -Wall`. Your change would basically move the issue to the cast from glad to passing the loader to glad (user code). 

Basically how it looks:

```c
typedef void (*GLADvoidfn)(void);
typedef GLADvoidfn (*GLADloadproc)(const char *name);
 
void* some_loader_func(const char *name) {
	return (void*)0;
}
 
void the_glad_loader(GLADloadproc load) {} 
 
int main(void) {
	the_glad_loader((GLADloadproc) some_loader_func);
	return 0;
}
```

The fix looks good to me, the only issue I can think of right now is with the builtin loader, but that can be 'fixed' by changing the signature to return the function pointer type -> good old `-pedantic` issue, but I guess that's fine.

**Comment by rajder** [comment]
Yes, void pointer to function pointer is the issue. I just added another problem in my last confused message by trying to skip the wrapper function for the loader.

My last message was initiated by me being confused by clang on OS X and looking for alternative solutions, it required "-Wpedantic" in the ignore diagnostics pragma. I was missing the "W" prefix which was not required by GCC.

As far as I can understand the C standard, a wrapper function must be used to avoid implementation defined behavior, if using my proposed change, even if the cast is accepted by the compiler.

I think I will try some more compilers before making a pull request. BR /Anders

**Comment by Dav1dde** [comment]
Pull Request -> #56

**Comment by grrrwaaa** [comment]
Related; found a couple of places where a pedantic compiler wouldn't convert void * to HMODULE:

`FreeLibrary(handle);`
to
`FreeLibrary((HMODULE)handle);`

`return GetProcAddress(handle, name);`
to
`return GetProcAddress((HMODULE)handle, name);`

**Comment by 0x8000-0000** [comment]
The problem is that in some build systems it is hard to tailor warning level per translation unit.

There are no problems with aliasing in this context - this is supported C/C++ on all platforms where dlsym and GetProcAddress are supported.

**Comment by Dav1dde** [comment]
The `glad2` branch now supports -pedantic (and all the other flags).

Support is achieved almost fully by using function pointers instead of `void*`, currently only in two places casts from function pointers to `void*` (or the other way round) are necessary:

1) within the utility loader function which stores the user loader function within the `void*` userpointer
2) within the internal loader (`dlsym`)

To still be able to compile with -pedantic all remaining casts are marked with
[`__extension__`](https://gcc.gnu.org/onlinedocs/gcc/Alternate-Keywords.html#Alternate-Keywords
) for `__GNUC__` compilers (e.g. GCC and clang).
